src/script/api/script_timemode.hpp

```diff
@@ -0,0 +1,51 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file script_timemode.hpp Switch the time mode. */
+
+#ifndef SCRIPT_TIMEMODE_HPP
+#define SCRIPT_TIMEMODE_HPP
+
+#include "script_object.hpp"
+
+/**
+ * Class to switch the current time.
+ * If you create an instance of this class, the mode will be switched to either calendar time or economy time mode.
+ * @note Destroying this object will restore the previous time mode.
+ * @api ai game
+ */
+class ScriptTimeMode : public ScriptObject {
+private:
+	bool last_time_mode; ///< The last time mode we were using.
+public:
+	/**
+	 * Creating an instance of this class switches the time mode used for queries and commands.
+	 * @param Calendar Should we use calendar time mode? (Set to false for economy time mode.)
+	 */
+	ScriptTimeMode(bool calendar);
+
+	/**
+	 * Destroying this instance resets the time mode to the mode it was in when the instance was created.
+	 */
+	~ScriptTimeMode();
+
+	/**
+	 * Set if the script is running in calendar time or economy time mode.
+	 * Calendar time is used by OpenTTD for technology like vehicle introductions and expiration, and variable snowline. It can be sped up or slowed down by the player.
+	 * Economy time always runs at the same pace and handles things like cargo production, everything related to money, etc.
+	 * @param Calendar Should we use calendar time mode? (Set to false for economy time mode.)
+	 */
+	static void SetTimeMode(bool calendar);
```

Comment:
Mode stuff is usually scoped (and authors are used to this by now I think), this means to change mode you open a new scope and create the object.
So `SetTimeMode()` is not the way to go as it implies authors need to store the old state themselves.

Reply:
Ah, so like async mode after all?

Reply:
Like any other mode actually.
To switch mode you create an instance of XXXMode class, and the previous mode is restored when instance is destroyed.
Exec and Test modes don't take parameters (the class name is explicit), while Company mode takes a company ID and Async takes the async state. But they all switch modes for as long the instance is in scope.
It's possible to do things like
```
/* do stuff as deity */
{
  local c0 = GSCompanyMode(0);
  /* do stuff as company 0 */
  {
    local c = GSCompanyMode(1);
    /* do stuff as company 1 */
  }
  /* do more stuff as company 0 */
  {
    local c = GSCompanyMode(2);
    /* do stuff as company 2 */
  }
}
/* do stuff as deity */
```

Reply:
Ah, I think I have that already. I just need to delete ScriptTimeMode::SetTimeMode() and ScriptTimeMode::IsCalendarTime() so script authors can use the object constructor/destructor instead?

Reply:
ScriptTimeMode::IsCalendarTime() will be needed for ScriptDate :)

Reply:
Okay, let me know if I got it right this time. 😉  Thanks for all your help with this.